File: backend/engines/audio_resolver.py
from dataclasses import dataclass


@dataclass
class AudioResolution:
    ok: bool
    song_id: str
    url: str = ""
    source: str = ""
    content_type: str = ""
    reason: str = ""

    @property
    def proxy_url(self) -> str:
        return f"/api/radio/audio/{self.song_id}" if self.ok and self.song_id else ""


class AudioResolver:
    def __init__(self, netease, store):
        self.netease = netease
        self.store = store

    async def resolve(self, song: dict, uid: str | None = None) -> AudioResolution:
        song_id = self._song_id(song)
        if not song_id:
            return AudioResolution(False, "", reason="missing_song_id")

        cached = await self.store.get_audio_resolution(song_id)
        if cached and cached.get("url"):
            return AudioResolution(
                True,
                song_id,
                cached["url"],
                cached.get("source", "cache"),
                cached.get("content_type", ""),
            )

        direct = await self._try_song_url(song_id, "song_url")
        if direct.ok:
            await self._cache(direct)
            return direct

        return AudioResolution(False, song_id, source="song_url", reason=direct.reason or "empty_url")

    async def resolve_with_candidates(
        self,
        song: dict,
        uid: str | None = None,
    ) -> AudioResolution:
        first = await self.resolve(song, uid=uid)
        if first.ok:
            return first

        query = self._query(song)
        if query:
            for candidate in await self.netease.search(query, limit=5):
                candidate_id = self._song_id(candidate)
                if not candidate_id or candidate_id == self._song_id(song):
                    continue
                candidate_resolution = await self._try_song_url(candidate_id, "search_candidate")
                if candidate_resolution.ok:
                    await self._cache(candidate_resolution)
                    return candidate_resolution

        await self.store.log_playback_event(
            "url_failed",
            song_id=self._song_id(song),
            uid=uid,
            reason=first.reason or "empty_url",
        )
        return first

    async def _try_song_url(self, song_id: str, source: str) -> AudioResolution:
        try:
            url = await self.netease.song_url(song_id)
        except Exception:
            return AudioResolution(False, song_id, source=source, reason="timeout")
        if not url:
            return AudioResolution(False, song_id, source=source, reason="empty_url")
        return AudioResolution(True, song_id, url, source, "audio/mpeg")

    async def _cache(self, resolved: AudioResolution) -> None:
        await self.store.save_audio_resolution(
            resolved.song_id,
            resolved.url,
            resolved.source,
            resolved.content_type,
        )

    def _song_id(self, song: dict | None) -> str:
        if not isinstance(song, dict):
            return ""
        value = song.get("id")
        return str(value).strip() if value is not None else ""

    def _artist_name(self, song: dict | None) -> str:
        if not isinstance(song, dict):
            return ""
        artist = song.get("artist")
        if isinstance(artist, str) and artist.strip():
            return artist.strip()
        for key in ("ar", "artists"):
            artists = song.get(key)
            if isinstance(artists, list) and artists:
                first = artists[0]
                if isinstance(first, dict):
                    name = first.get("name")
                    if isinstance(name, str):
                        return name.strip()
        return ""

    def _query(self, song: dict | None) -> str:
        if not isinstance(song, dict):
            return ""
        name = str(song.get("name") or "").strip()
        artist = self._artist_name(song)
        return f"{artist} {name}".strip()

File: src/services/audioResolver.ts
import type { Track } from "../types.js";
import { normalizeMatchText } from "../utils/text.js";
import type { MemoryStore } from "../storage/memoryStore.js";
import type { NeteaseService } from "./neteaseService.js";

export interface AudioResolution {
  ok: boolean;
  songId: string;
  url: string;
  source: string;
  contentType: string;
  reason: string;
  proxyUrl: string;
}

function result(args: Partial<AudioResolution> & { ok: boolean; songId: string }): AudioResolution {
  const ok = args.ok;
  const songId = args.songId;
  return {
    ok,
    songId,
    url: args.url || "",
    source: args.source || "",
    contentType: args.contentType || "",
    reason: args.reason || "",
    proxyUrl: ok && songId ? `/api/radio/audio/${songId}` : "",
  };
}

export class AudioResolver {
  constructor(
    private readonly netease: NeteaseService,
    private readonly store: MemoryStore,
  ) {}

  async resolve(track: Track, uid: string | null = null, forceRefresh = false): Promise<AudioResolution> {
    if (!track.id) return result({ ok: false, songId: "", reason: "missing_song_id" });

    const cached = forceRefresh ? null : this.store.getAudioResolution(track.id);
    if (cached?.url && this.isPlayableUrl(cached.url)) {
      return result({
        ok: true,
        songId: track.id,
        url: cached.url,
        source: cached.source || "cache",
        contentType: cached.contentType || "audio/mpeg",
      });
    }

    const direct = await this.trySongUrl(track.id, "song_url");
    if (direct.ok) {
      this.store.saveAudioResolution(direct.songId, direct.url, direct.source, direct.contentType);
      return direct;
    }

    this.store.logPlaybackEvent("url_failed", {
      uid,
      songId: track.id,
      reason: direct.reason || "empty_url",
    });
    return direct;
  }

  async resolveWithCandidates(track: Track, uid: string | null = null, forceRefresh = false): Promise<AudioResolution> {
    const first = await this.resolve(track, uid, forceRefresh);
    if (first.ok) return first;

    const query = [track.artist, track.name].filter(Boolean).join(" ").trim();
    if (!query) return first;

    const candidates = await this.netease.search(query, 5);
    for (const candidate of candidates) {
      if (!candidate.id || candidate.id === track.id) continue;
      if (!this.isSameRecording(track, candidate)) continue;
      const candidateResolution = await this.trySongUrl(candidate.id, "search_candidate");
      if (candidateResolution.ok) {
        this.store.saveAudioResolution(
          candidateResolution.songId,
          candidateResolution.url,
          candidateResolution.source,
          candidateResolution.contentType,
        );
        return candidateResolution;
      }
    }

    return first;
  }

  private async trySongUrl(songId: string, source: string): Promise<AudioResolution> {
    try {
      const url = await this.netease.songUrl(songId);
      if (!url) return result({ ok: false, songId, source, reason: "empty_url" });
      if (!this.isPlayableUrl(url)) return result({ ok: false, songId, source, reason: "unplayable_url" });
      return result({ ok: true, songId, source, url, contentType: "audio/mpeg" });
    } catch {
      return result({ ok: false, songId, source, reason: "timeout" });
    }
  }

  private isPlayableUrl(url: string): boolean {
    return Boolean(url) && !url.includes("music.163.com/song/media/outer/url");
  }

  private isSameRecording(left: Track, right: Track): boolean {
    const leftName = normalizeMatchText(left.name);
    const rightName = normalizeMatchText(right.name);
    if (!leftName || leftName !== rightName) return false;
    const leftArtist = normalizeMatchText(left.artist);
    const rightArtist = normalizeMatchText(right.artist);
    return !leftArtist || leftArtist === rightArtist;
  }
}

